core/market/src/market/agreement.rs

```diff
@@ -13,10 +9,50 @@ use crate::db::DbMixedExecutor;
 
 pub async fn bind_gsb(db: DbMixedExecutor, public_prefix: &str, _local_prefix: &str) {
     log::trace!("Binding market agreement public service to service bus");
-    ServiceBinder::new(public_prefix, &db, ()).bind(get_agreement);
+    ServiceBinder::new(public_prefix, &db, ())
+        .bind(list_agreements)
+        .bind(get_agreement);
     log::debug!("Successfully bound market agreement public service to service bus");
 }
 
+async fn list_agreements(
+    db: DbMixedExecutor,
+    _sender_id: String,
+    msg: ListAgreements,
+) -> Result<Vec<AgreementListEntry>, RpcMessageError> {
+    let dao = db.as_dao::<AgreementDao>();
+
+    let agreements = dao
+        .list(
+            None,
+            msg.state.map(Into::into),
+            msg.before_date,
+            msg.after_date,
+            msg.app_session_id,
+        )
+        .await
+        .map_err(|e| RpcMessageError::Market(e.to_string()))?;
+
+    let mut result = Vec::new();
+    let naive_to_utc = |ts| DateTime::<Utc>::from_utc(ts, Utc);
+
+    for agreement in agreements {
+        let role = match agreement.id.owner() {
+            Owner::Provider => Role::Provider,
+            Owner::Requestor => Role::Requestor,
+        };
+
+        result.push(AgreementListEntry {
+            id: agreement.id.into_client(),
+            creation_ts: naive_to_utc(agreement.creation_ts),
+            approve_ts: agreement.approved_ts.map(naive_to_utc),
+            role: role,
+        });
+    }
+
+    Ok(result)
+}
+
```

Comment:
You could use the second `list_agreements` function internally and pass parameters from `msg`

Reply:
Not exactly, the other `list_agreements` requires a reference to MarketService, which can't be constructed just from `DbMixedExecutor`.

I don't like the duplicated logic either, but I don't see a clear way to extract it. You've done the same with your `get_agreement` functions, too.

Reply:
`ServiceBinder::bind_with_processor` can pass additional generic parameter to bound functions. And `agreement::bind_gsb` could have `self` easily.

Anyway, approving